### ISSUE TITLE
Not obviously possible to see scalac help in sbt

### ISSUE THREAD

**Issue by tpolecat** [sbt/sbt#3123]
So, I only use sbt and I haven't had scalac installed for years. This means I can't say `scalac -Xlint:help` to see the help options. I end up looking at scalac source which is ridiculous.

So, it would be nice if there were an easy way to get these help messages inside sbt. 

It may be trivial but it's not obvious to me. I can `set scalacOptions += "-Xlint:help"` and then `console` and scroll back above the stacktrace and read the help and then filter it back out of `scalacOptions` but that's obviously dumb.

Any easy win here?

**Comment by tpolecat** [comment]
x-ref https://github.com/scala/scala/pull/5402#issuecomment-295333016

**Comment by eed3si9n** [comment]
Maybe an input task that works like running `scala` Bash script from the terminal?

```
> scala -Y
Usage: scala <options> <source files>...
> scala
## probably internally call `console` task here
Welcome to Scala version 2.12.2 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_91).
```

**Comment by fommil** [comment]
this is starting to sound a bit like the `ensimeCompileOnly` task, which I was hoping to allow to take extra params... and that would be generally useful for compiling one file too.

**Comment by tpolecat** [comment]
Yeah, that's not a bad starting point. Just bare scala with nothing from the build. I would be ok with that although it might end up being confusing and would attract feature requests like "add my project dependencies" and so on. Not sure what to do about that.

**Comment by fommil** [comment]
https://github.com/ensime/ensime-sbt/blob/2.0/src/main/scala/EnsimeExtrasPlugin.scala#L172-L195

imagine this could also parse inputs so the user can say

```
> foo/ensimeCompileOnly -Xlint:help

> foo/ensimeCompileOnly -Xmagic-thing /path/to/file
```

and, far superior if this was `compileOnly` (or `scalac`) built in to `sbt`, not just an ensime thing. The nice thing in Emacs is that the file path can be inserted automatically.

**Comment by eed3si9n** [comment]
If we delegate `scala` without input parameter to `console`, then it would actually include my project dependencies. I'd expect passing source files to work the same.

**Comment by tpolecat** [comment]
> If we delegate scala without input parameter to console, then it would actually include my project dependencies. I'd expect passing source files to work the same.

Right but then you have to decide what other stuff to pass right? Like, compiler plugins, scalacOptions (in which config?) etc. Seems like a recipe for confusion if there's `console` and `scala` that are superficially similar but not identical; or `scala` that behaves totally differently without args.

**Comment by eed3si9n** [comment]
Also the (input) task name `scala` gets confusing when we try to talk about it, so I'm open to suggestions. Some candidates:

- `scalaHelp`: don't accept `<file>` args, and require all args start with `-`. (defaults to `-help`)
- `compileOnly`: accepts `<file>` args and `-` options. (classpath is respected?)

**Comment by fommil** [comment]
@eed3si9n `compileOnly` uses exactly what `compile` uses, so `scalacOptions`, `classpath`, target output, etc, etc. It even has its own namespace so you can have extra (or less) flags for that target. I should have added the parser, it'd be so useful. e.g. to be able to see a file at various stages in the compile cycle (sometimes I want to see pre, sometimes post, typer)

**Comment by fommil** [comment]
option 1 should probably be `scalacHelp` (note the `c`) but my preference is for 2! You're welcome to lift our impl and tests wholesale from `sbt-ensime`, it's all Apache 2.0 (just needs a simple parser).

**Comment by som-snytt** [comment]
I have bash aliases that add options like `scalac -feature -deprecation` and `scala -Dscala.color`. Every so often, I want to test in isolation and I'm momentarily confused and annoyed because I forgot that I'm running under `-deprecation`. Similarly, if I'm running under sbt with various options set, sometimes I just want to run a compile with `-Xfuture` or `-Xlint:help` or what have you. `> compile -Xlint:help` seems unambiguous to me.

**Comment by dwijnand** [comment]
Perhaps we should steal the scala/scalac/scaladoc/scalap commands in the scala/scala build and provide them out the box in sbt.

**Comment by som-snytt** [comment]
In REPL:
```
scala> :se -Xlint:help

scala> val cmd = new scala.tools.nsc.CompilerCommand(Nil, $intp.settings)
cmd: scala.tools.nsc.CompilerCommand = scala.tools.nsc.CompilerCommand@5d497a91

scala> cmd.getInfoMessage($intp.global)
res0: String =
"Enable or disable specific warnings
  adapted-args               Warn if an argument list is modified to match the receiver.
  nullary-unit               Warn when nullary methods return Unit.
  inaccessible               Warn about inaccessible types in method signatures.
  nullary-override           Warn when non-nullary `def f()' overrides nullary `def f'.
  infer-any                  Warn when a type argument is inferred to be `Any`.
  missing-interpolator       A string literal appears to be missing an interpolator id.
  doc-detached               A Scaladoc comment appears to be detached from its element.
  private-shadow             A private field (or class parameter) shadows a superclass field.
  type-parameter-shadow      A local type ...
```
